Arduino_related_code/ValveCaliberationTest.py

```diff
@@ -0,0 +1,52 @@
+import tkinter as tk
+from threading import Thread
+from Arduino_related_code.ArduinoDigital import ArduinoDigital
+from Arduino_related_code.ValveControl import ValveControl
+import time
+
+def calibrate_valve(pin, duration_ms, arduino):
+    valve_control = ValveControl(pin, arduino)
+    duration_s = duration_ms / 1000  # Convert duration from milliseconds to seconds
+
+    for i in range(25):
+        print(f"Calibrating valve on pin {pin}: Iteration {i+1}")
+        valve_control.OpenValve(duration_s)  # Open valve for the specified duration in seconds
+
+        while valve_control.IsValveOpen():
+            time.sleep(0.1)  # Check every 0.1 seconds
```

Comment:
The sleep here is messing up your valve timing  calculations. you are inserting a 100mSec inaccuracy in your measurements. The IsValveOpen() function will close the valve if the time is up. So calling it every 0.1 of a second causes an inherit inaccuracy in measuring the valve opening during. Get rid of the sleep. Just do a loop until IsValveOpen returns true.

I think we need to move the valve timing control to Arduino. I'll open an issue for this.

Reply:
I agree the sleep will lead to inaccuracies.Thanks for pointing it out.I will remove it